### PR TITLE
Small upgrade Observation

### PR DESCRIPTION
**Pull request by DeadLineSS13**
## About the PR
RAISE_EVENT worked faster then GLOB.thing.raise_event. And RAISE_EVENT can return value.

## Changelog
🆑 DeadLineSS13
refactor: Observation small refactored.
/🆑

### CODE REVIEW

File: code/datums/datum.dm [hunk 1]
```diff
@@ -9,6 +9,15 @@
 	/// Datum level flags
 	var/datum_flags = NONE
 
+	/**
+	  * Any datum registered to receive observations from this datum is in this list
+	  *
+	  * Lazy associated list in the structure of `observation:registree/list of registrees`
+	  */
+	var/list/observations
```
Comment (Kmc2000) [inline code/datums/datum.dm:17]: You have an uninitialised list here, I’m assuming it’s initialised elsewhere?

May be useful to include a wrapper to do this easily as you do with RaiseEvent

File: code/datums/datum.dm [hunk 2]
```diff
@@ -9,6 +9,15 @@
...
+	/// Lazy associated list in the structure of `observations:proctype` that are run when the datum receives that observation
+	var/list/list/datum/callback/observation_procs
```
Comment (Kmc2000) [inline code/datums/datum.dm:19]: Strictly typing this list may lead to wackiness, I’m not 100% sure on how well BYOND actually deals with these things. I believe the convention is to use a loosely typed assoc list with type checking as needed.

Comment (Kmc2000) [reply code/datums/datum.dm:19]: Again, if not already existent, a wrapper proc to add an event listener could be nice here.

**Review by Kmc2000 — APPROVED** [review]
Pretty clever way of doing this. Seems to be a very stripped down version of signals, which I’m not sure bay has. In that case though, consider porting them over, it’s basically this but somewhat easier to use.

I’d say this looks fine at a glance, though I’m not 100% familiar with this codebase’s way of handling events.

**Comment by DeadLineSS13** [comment]
> Pretty clever way of doing this. Seems to be a very stripped down version of signals, which I’m not sure bay has. In that case though, consider porting them over, it’s basically this but somewhat easier to use.
> 
> 
> 
> I’d say this looks fine at a glance, though I’m not 100% familiar with this codebase’s way of handling events.

Observation is just a version of Signals Bay. The problem is that all the events were stored globally and the raise_event response was much longer in the end. I decided to attach this moment to datums (that is, as on TG). But at the same time, it is much better than Signals TG, since data on what events there are can be viewed through GLOB. And at the same time, you can change the register/unregister/raise_event for each event. That is, it is more flexible and more lucid than Signals.

File: code/datums/datum.dm (hunk 2 above)
Comment (DeadLineSS13) [reply code/datums/datum.dm:19]: Byond have dynamic typization.
Example create var/thing = list()
thing will contain list, but don't have memory access for work with list (thing.len, thing.Copy, thing.Cut and etc)

**Pull request closed without merging**